R/parse_people.R

```diff
@@ -1,98 +1,164 @@
-
 #' @importFrom utils as.person
 #' @importFrom methods is
 parse_people <- function(people, codemeta) {
-  if (!is(people, "person")) {
-    people <- as.person(people)
-  }
+
+  people <- as.person(people)
+
   if (length(people) == 0) {
+
     return(codemeta)
   }
 
   ## people with no role are assumed to be "Author" role
-  codemeta$author <-
-    c(people_with_role(people, "aut"),
-      people_without_role(people))
-  codemeta$contributor <- c(people_with_role(people, "ctb"),
-                            people_with_role(people, "com"),
-                            people_with_role(people, "dtc"),
-                            people_with_role(people, "ths"),
-                            people_with_role(people, "trl"))
-  codemeta$copyrightHolder <- people_with_role(people, "cph")
-  codemeta$funder <- people_with_role(people, "fnd")
-  codemeta$maintainer <- people_with_role(people, "cre")
-  codemeta
-}
+  role_mapping <- list(
+    author = c("aut", NA), # NA = people without a role
+    contributor = c("ctb", "com", "dtc", "ths", "trl"),
+    copyrightHolder = "cph",
+    funder = "fnd",
+    maintainer = "cre"
+  )
+
+  # get the names of the role entries in codemeta as given in role_mapping
+  roles <- names(role_mapping)
 
-people_without_role <- function(people) {
-  index <- vapply(people, function(p)
-    is.null(p$role), logical(1))
-  lapply(people[index], person_to_schema)
+  # call people_with_role on each element of role_mapping and assign the
+  # resulting lists to the corresponding entries in codemeta
+  codemeta[roles] <- lapply(role_mapping, people_with_role, people = people)
+
+  # return codemeta
+  codemeta
 }
 
+# people_with_role -------------------------------------------------------------
 people_with_role <- function(people, role = "aut") {
+
+  # If there is more than one role requested, call this function recursively
+  # for each role, combine the results with c() and return
+  if (length(role) > 1) {
+
+    return(do.call(c, lapply(role, people_with_role, people = people)))
+  }
+
+  # if role is NA, has_role is TRUE for all people without any role!
   has_role <- locate_role(people, role)
-  if (any(has_role)) {
-    out <- lapply(people[has_role], person_to_schema)
-  } else {
-    out <- NULL
+
+  # return NULL if there are no people with the required (or without any) role
+  if (! any(has_role)) {
+
+    return(NULL)
   }
-  out
+
+  # create schema for each person with the selected (or without any) role
+  lapply(people[has_role], person_to_schema)
 }
 
+# locate_role ------------------------------------------------------------------
+# role = NA returns TRUE for all people without a role
+#' @importFrom purrr map_lgl
 locate_role <- function(people, role = "aut") {
-  vapply(people, function(p)
-    any(grepl(role, p$role)), logical(1))
+
+  has_no_role <- function(p) is.null(p$role)
+  has_role <- function(p, role) any(grepl(role, p$role))
```

Comment:
could you define `has_no_role()` and `has_role()` outside of `locate_role()`? Thanks.

Reply:
Of course. They are now outside, renamed to `person_has_no_role()` and `person_has_role()`, respectievly, and tested.